timelock-server/src/main/java/com/palantir/atlasdb/timelock/lock/ImmutableTimestampTracker.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock.lock;
+
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.UUID;
+
+import com.google.common.collect.Maps;
+import com.palantir.atlasdb.timelock.util.LoggableIllegalStateException;
+import com.palantir.logsafe.SafeArg;
+
+import net.jcip.annotations.GuardedBy;
+
+public class ImmutableTimestampTracker {
+
+    @GuardedBy("this")
+    private final SortedMap<Long, UUID> holdersByTimestamp = Maps.newTreeMap();
+
+    public synchronized void lock(long timestamp, UUID requestId) {
```

Comment:
I ended up moving all the locking logic into this class (and adding error messages that mention "locking"), so I felt it made sense to go ahead and call these methods `lock` and `unlock`. Maybe I should rename this class...